File: qr-gif-generator.m
#import <Cocoa/Cocoa.h>
#import <CoreImage/CoreImage.h>
#import <ImageIO/ImageIO.h>
#import <CoreServices/CoreServices.h>

// Helper to generate scaled QR Code CGImage
CGImageRef createScaledQRImage(NSString *text) {
    NSData *data = [text dataUsingEncoding:NSUTF8StringEncoding];
    CIFilter *filter = [CIFilter filterWithName:@"CIQRCodeGenerator"];
    [filter setValue:data forKey:@"inputMessage"];
    [filter setValue:@"M" forKey:@"inputCorrectionLevel"]; // Medium error correction
    
    CIImage *ciImage = filter.outputImage;
    if (!ciImage) return nil;
    
    // Scale the image up (CIQRCodeGenerator outputs a tiny image)
    CGFloat scaleX = 450.0 / ciImage.extent.size.width;
    CGFloat scaleY = 450.0 / ciImage.extent.size.height;
    CIImage *scaledImage = [ciImage imageByApplyingTransform:CGAffineTransformMakeScale(scaleX, scaleY)];
    
    CIContext *context = [CIContext contextWithOptions:nil];
    return [context createCGImage:scaledImage fromRect:scaledImage.extent];
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *text = @"";
        
        // 1. Check arguments
        if (argc > 1) {
            NSMutableArray *args = [NSMutableArray array];
            for (int i = 1; i < argc; i++) {
                [args addObject:[NSString stringWithUTF8String:argv[i]]];
            }
            text = [args componentsJoinedByString:@" "];
        } else {
            // 2. Check stdin
            if (!isatty(STDIN_FILENO)) {
                NSData *inputData = [[NSFileHandle fileHandleWithStandardInput] readDataToEndOfFile];
                text = [[NSString alloc] initWithData:inputData encoding:NSUTF8StringEncoding];
            }
            
            // 3. Check clipboard if still empty
            if (!text || [text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]].length == 0) {
                NSPasteboard *pc = [NSPasteboard generalPasteboard];
                text = [pc stringForType:NSPasteboardTypeString];
            }
        }
        
        text = [text stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceAndNewlineCharacterSet]];
        
        if (!text || text.length == 0) {
            fprintf(stderr, "Error: Input text is empty. Please copy some text to clipboard, pass it as an argument, or pipe it to stdin.\n");
            return 1;
        }
        
        // Chunk the text into parts of 100 characters (low density = very easy to scan)
        NSUInteger chunkSize = 100;
        NSMutableArray *chunks = [NSMutableArray array];
        NSUInteger length = text.length;
        for (NSUInteger i = 0; i < length; i += chunkSize) {
            NSRange range = NSMakeRange(i, MIN(chunkSize, length - i));
            [chunks addObject:[text substringWithRange:range]];
        }
        
        NSUInteger total = chunks.count;
        NSMutableArray *images = [NSMutableArray array];
        
        for (NSUInteger i = 0; i < total; i++) {
            NSString *formatted = [NSString stringWithFormat:@"%lu/%lu:%@", i + 1, total, chunks[i]];
            CGImageRef img = createScaledQRImage(formatted);
            if (img) {
                [images addObject:(__bridge id)img];
                CGImageRelease(img);
            } else {
                fprintf(stderr, "Error: Failed to generate QR code for chunk %lu.\n", i + 1);
                return 1;
            }
        }
        
        NSString *tempDir = NSTemporaryDirectory();
        NSString *outputPath = [tempDir stringByAppendingPathComponent:@"qr_transfer.gif"];
        NSURL *outputURL = [NSURL fileURLWithPath:outputPath];
        
        CGImageDestinationRef destination = CGImageDestinationCreateWithURL((__bridge CFURLRef)outputURL, kUTTypeGIF, images.count, nil);
        if (!destination) {
            fprintf(stderr, "Error: Failed to create image destination.\n");
            return 1;
        }
        
        // Loop count of 0 = loop forever
        NSDictionary *fileProperties = @{
            (NSString *)kCGImagePropertyGIFDictionary: @{
                (NSString *)kCGImagePropertyGIFLoopCount: @0
            }
        };
        CGImageDestinationSetProperties(destination, (__bridge CFDictionaryRef)fileProperties);
        
        // Delay: 0.25 seconds per frame
        NSDictionary *frameProperties = @{
            (NSString *)kCGImagePropertyGIFDictionary: @{
                (NSString *)kCGImagePropertyGIFDelayTime: @0.25
            }
        };
        
        for (id imgObj in images) {
            CGImageRef img = (__bridge CGImageRef)imgObj;
            CGImageDestinationAddImage(destination, img, (__bridge CFDictionaryRef)frameProperties);
        }
        
        if (CGImageDestinationFinalize(destination)) {
            printf("%s\n", [outputPath UTF8String]);
        } else {
            fprintf(stderr, "Error: Failed to finalize GIF.\n");
            CFRelease(destination);
            return 1;
        }
        
        CFRelease(destination);
    }
    return 0;
}

File: qr-gif-generator.swift
import Cocoa
import CoreImage
import ImageIO
import CoreServices

// Helper to read stdin
func readStdin() -> String? {
    var data = Data()
    while let line = readLine(strippingNewline: false) {
        if let lineData = line.data(using: .utf8) {
            data.append(lineData)
        }
    }
    return String(data: data, encoding: .utf8)
}

func generateQRCode(from string: String) -> CGImage? {
    guard let data = string.data(using: .utf8) else { return nil }
    guard let filter = CIFilter(name: "CIQRCodeGenerator") else { return nil }
    filter.setValue(data, forKey: "inputMessage")
    filter.setValue("M", forKey: "inputCorrectionLevel") // Medium error correction
    guard let ciImage = filter.outputImage else { return nil }
    
    // Scale the image up (CIQRCodeGenerator outputs a tiny image, around 27x27 pixels)
    // We scale to 450x450 to make it large and easy to scan
    let scaleX = 450.0 / ciImage.extent.width
    let scaleY = 450.0 / ciImage.extent.height
    
    // Use CGAffineTransform for scaling
    let scaledImage = ciImage.transformed(by: CGAffineTransform(scaleX: scaleX, y: scaleY))
    
    let context = CIContext(options: [CIContextOption.useSoftwareRenderer: false])
    return context.createCGImage(scaledImage, from: scaledImage.extent)
}

func createGIF(from images: [CGImage], delay: Double, destinationURL: URL) -> Bool {
    guard let destination = CGImageDestinationCreateWithURL(destinationURL as CFURL, kUTTypeGIF, images.count, nil) else {
        return false
    }
    
    // Loop count of 0 makes the GIF loop forever
    let fileProperties = [kCGImagePropertyGIFDictionary as String: [kCGImagePropertyGIFLoopCount as String: 0]]
    CGImageDestinationSetProperties(destination, fileProperties as CFDictionary)
    
    let frameProperties = [kCGImagePropertyGIFDictionary as String: [kCGImagePropertyGIFDelayTime as String: delay]]
    
    for image in images {
        CGImageDestinationAddImage(destination, image, frameProperties as CFDictionary)
    }
    
    return CGImageDestinationFinalize(destination)
}

func main() {
    var text = ""
    
    // 1. Check arguments
    if CommandLine.arguments.count > 1 {
        text = CommandLine.arguments.dropFirst().joined(separator: " ")
    } else {
        // 2. Check stdin
        let isPipable = !isatty(STDIN_FILENO)
        if isPipable, let stdinText = readStdin(), !stdinText.isEmpty {
            text = stdinText
        } else {
            // 3. Check clipboard
            if let clipboardText = NSPasteboard.general.string(forType: .string), !clipboardText.isEmpty {
                text = clipboardText
            }
        }
    }
    
    text = text.trimmingCharacters(in: .whitespacesAndNewlines)
    
    if text.isEmpty {
        fputs("Error: Input text is empty. Please copy some text to clipboard, pass it as an argument, or pipe it to stdin.\n", stderr)
        exit(1)
    }
    
    // Chunk size: 250 characters ensures a very low density QR code that scans instantly
    let chunkSize = 250
    var chunks: [String] = []
    
    let chars = Array(text)
    var index = 0
    while index < chars.count {
        let endIndex = min(index + chunkSize, chars.count)
        let chunkPayload = String(chars[index..<endIndex])
        chunks.append(chunkPayload)
        index += chunkSize
    }
    
    let total = chunks.count
    var qrImages: [CGImage] = []
    
    for (i, payload) in chunks.enumerated() {
        // Format: index/total:payload
        let formattedText = "\(i + 1)/\(total):\(payload)"
        if let cgImage = generateQRCode(from: formattedText) {
            qrImages.append(cgImage)
        } else {
            fputs("Error: Failed to generate QR code for chunk \(i + 1).\n", stderr)
            exit(1)
        }
    }
    
    let tempURL = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent("qr_transfer.gif")
    // Delay: 0.25 seconds per frame
    let success = createGIF(from: qrImages, delay: 0.25, destinationURL: tempURL)
    
    if success {
        print(tempURL.path)
    } else {
        fputs("Error: Failed to write animated GIF.\n", stderr)
        exit(1)
    }
}

main()
